src/asl.ts

```diff
@@ -5526,28 +5253,1768 @@ export namespace ASLGraph {
+          return extractVariableReferences(parameters).map((x) => ({
+            type: "FilterPropAssignment",
+            from: x,
+            to: resultPath,
+            props,
+          }));
+        }
+      }
+      // if the value is json path, it will not be a boolean or string
+      return [
+        resultPath === "[[task]]" || resultPath === "[[map]]"
+          ? {
+              type: "StateInputProps",
+              from: parameters as string,
+              props,
+              stateType: resultPath === "[[task]]" ? "Task" : "Map",
+            }
+          : {
+              type: "PropAssignment",
+              from: parameters as string,
+              to: resultPath,
+              props,
+            },
+      ];
+    }
+
+    /**
+     * Pull variable references out of intrinsic functions and json path filters.
+     *
+     * States.Array($.var) => $.var
+     * $.arr[?(['value'] == $.var)] => $.arr, $.var
+     */
+    function extractVariableReferences(jsonPath: string) {
+      return jsonPath.match(/(\$\.?[^,)]+)+/g) ?? [];
+    }
+  }
+
+  function getVariableNamesFromState(state: State): string[] {
+    const names = ((state) => {
+      if (
+        isTaskState(state) ||
+        isMapTaskState(state) ||
+        isParallelTaskState(state)
+      ) {
+        return [
+          state.ResultPath,
+          ...(state.Catch?.map((c) => c.ResultPath) ?? []),
+        ];
+      } else if (isPassState(state)) {
+        return [state.ResultPath];
+      } else if (
+        isSucceedState(state) ||
+        isFailState(state) ||
+        isChoiceState(state) ||
+        isWaitState(state)
+      ) {
+        return [];
+      }
+      return assertNever(state);
+    })(state);
+
+    return names.filter((n): n is string => !!n);
+  }
+
+  export const DefaultOptimizeOptions: OptimizeOptions = {
+    optimizeVariableAssignments: true,
+    removeUnreachableStates: true,
+    joinConsecutiveChoices: true,
+    removeNoOpStates: true,
+  };
+
+  export interface OptimizeOptions {
+    /**
+     * Attempts to collapse assignments that provide no functional value.
+     *
+     * Collapses assignments created by the interpreter and the input source.
+     *
+     * ```ts
+     * const a = "a"
+     * const b = a;
+     * =>
+     * const b = "a"
+     * ```
+     */
+    optimizeVariableAssignments: boolean;
+    /**
+     * Traverses the graph transitions and removes any states that cannot be reached.
+     * Unreachable states will fail on deployment to AWS Step Functions.
+     *
+     * Note: If this optimization is turned off, it is likely the output ASL will fail.
+     *       Functionless creates some unreachable that this logic is expected to remove.
+     *       Only turn this off if you intend to preform similar logic or use the ASL for another
+     *       purpose than AWS Step Functions.
+     */
+    removeUnreachableStates: boolean;
```

Comment:
Why would we ever turn this off? Should it even be an option?

Reply:
In case someone wanted to implement a different algorithm or if we want to use it for debugging. Doesn't hurt to have the flag.